kafka-impl/src/main/java/io/streamnative/pulsar/handlers/kop/coordinator/group/OffsetAcker.java

```diff
@@ -13,213 +13,134 @@
-                .receiverQueueSize(0)
-                .subscriptionInitialPosition(SubscriptionInitialPosition.Earliest);
-        this.brokerService = brokerService;
-    }
-
-    public void addOffsetsTracker(String groupId, byte[] assignment) {
-        ByteBuffer assignBuffer = ByteBuffer.wrap(assignment);
-        Assignment assign = ConsumerProtocol.deserializeAssignment(assignBuffer);
-        if (log.isDebugEnabled()) {
-            log.debug(" Add offsets after sync group: {}", assign.toString());
+    public CompletableFuture<Map<TopicPartition, Errors>> ackOffsets(
+            final String groupId, final Map<TopicPartition, OffsetAndMetadata> partitionAndOffsets) {
+        if (groupId == null || groupId.isEmpty()) {
+            log.warn("Try to ackOffsets for group {}", groupId);
+            return CompletableFuture.completedFuture(
+                    CoreUtils.mapValue(partitionAndOffsets, offsetAndMetadata -> Errors.INVALID_GROUP_ID));
         }
-        assign.partitions().forEach(topicPartition -> getOrCreateConsumer(groupId, topicPartition));
-    }
 
-    public void ackOffsets(String groupId, Map<TopicPartition, OffsetAndMetadata> offsetMetadata) {
-        if (log.isDebugEnabled()) {
-            log.debug(" ack offsets after commit offset for group: {}", groupId);
-            offsetMetadata.forEach((partition, metadata) ->
-                log.debug("\t partition: {}",
-                    partition));
-        }
-        offsetMetadata.forEach(((topicPartition, offsetAndMetadata) -> {
-            // 1. get consumer, then do ackCumulative
-            CompletableFuture<Consumer<byte[]>> consumerFuture = getOrCreateConsumer(groupId, topicPartition);
+        final CompletableFuture<Map<TopicPartition, Errors>> future = new CompletableFuture<>();
+        final Map<TopicPartition, Errors> ackResultMap = new ConcurrentHashMap<>();
+        final int numPartitions = partitionAndOffsets.size();
 
-            consumerFuture.whenComplete((consumer, throwable) -> {
-                if (throwable != null) {
-                    log.warn("Failed to create offset consumer for [group={}] [topic={}]: {}",
-                            groupId, topicPartition, throwable.getMessage());
-                    removeConsumer(groupId, topicPartition);
+        final BiConsumer<TopicPartition, Errors> addAckResult = (topicPartition, errors) -> {
+            ackResultMap.put(topicPartition, errors);
+            if (ackResultMap.size() == numPartitions) {
+                future.complete(ackResultMap);
+            }
+        };
+
+        partitionAndOffsets.forEach((topicPartition, offsetAndMetadata) -> {
+            final String partitionTopicName = new KopTopic(topicPartition.topic())
+                    .getPartitionName(topicPartition.partition());
+            // 1. find PersistentTopic
+            brokerService.getTopicIfExists(partitionTopicName).whenComplete((optTopic, e) -> {
```

Comment:
Why not use `admin.topics().resetCursor` api to commit offset, which can avoid topic ownership change.

Reply:
Because even if we use `PulsarAdmin` to reset cursor, we still need to `getTopic` in current broker so that we can find the `Position` of the offset.

Reply:
Another reason is, when Kafka handles `OFFSET_COMMIT` request, it will also eventually call `ReplicaManager#appendRecords`. `ReplicaManager#getPartitionAndLeaderReplicaIfLocal` is responsible to check if the current broker is the leader of the partition. If not, a `NotLeaderForPartitionException` will be thrown and converted to the `NOT_LEADER_FOR_PARTITIONE` error code later.

Reply:
OK